Envelope agnostic binary protocol

kriskowal opened the pull request:
This adds a DecodeRequest method to the Binary protocol, making it capable of
automatically inferring whether a request buffer contains an envelope and
decoding the body regardless.

It goes on to add a RequestProtocol interface to extend the exisiting Protocol
interface, including DecodeRequest.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/thriftrw/thriftrw-go?pullRequest=339) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/thriftrw/thriftrw-go?pullRequest=339) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/thriftrw/thriftrw-go?pullRequest=339) it.</sub>

breerly commented on protocol/binary.go:
```diff
@@ -61,3 +61,30 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
 	return e, err
 }
+
+func (b binaryProtocol) DecodeRequest(r io.ReaderAt) (wire.Value, error) {
```
This probably deserves a nice description an potentially a background into why this is useful/necessary :)

kriskowal replied:
So it is done.

kriskowal commented:
I’ve added changes…did not modify the type of Binary, added EnvelopeAgnosticBinary instead. Added a return enum to DecodeRequest to indicate what kind of envelope was around the request so the RPC layer can respond in kind.

prashantv commented on protocol/binary.go:
```diff
@@ -61,3 +67,53 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
+		return e.Value, VersionedMessage, nil
+	}
+
+	// All other patterns are either bare structs or invalid.
```
rule 4 was,
```
	// 4. A message of length >1 starting with 0x00-0x0f can only be the
	// beginning of an un-enveloped field.
```

However, we're assuming here that anything where `buf[0] != 0` must be a non-enveloped struct. Isn't it possible it's an enveloped non-strict value with a long string as the name?

If we're ignoring that case, then it seems like we should very explicitly call it out.

kriskowal commented on protocol/binary.go:
```diff
@@ -61,3 +67,53 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
 	return e, err
 }
+
+// DecodeRequest specializes Decode and replaces DecodeEnveloped for the
```
I’ve clarified that the request is a struct. Requests and responses are always implicit TStructs on the wire. There is nowhere in the protocol to specify otherwise.

kriskowal commented on protocol/binary_test.go:
```diff
@@ -678,21 +696,32 @@ func TestBinaryEnvelopeErrors(t *testing.T) {
 	for _, tt := range tests {
 		reader := bytes.NewReader(tt.encoded)
-		_, err := Binary.DecodeEnveloped(reader)
+		_, err := EnvelopeAgnosticBinary.DecodeEnveloped(reader)
```
Need not, should not, fixed.

kriskowal commented on protocol/binary_test.go:
```diff
@@ -776,6 +808,24 @@ func TestBinaryEnvelopeSuccessful(t *testing.T) {
+		}
+
+		// Verify a round trip, back from encode after decode.
+
```
I’d like to keep the blank. A comment without a blank I infer applies only to the immediate following block, a comment with a blank applies until the next comment with a following blank. In this case, this comment applies to the remaining two checks, which have a blank between them.

breerly replied:
Makes me wonder if those sections should be encapsulated into a single method/func... likely thinking about it too hard.

prashantv commented on protocol/binary.go:
```diff
@@ -61,3 +82,108 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
+	// 1. A message of length 1 containing only 0x00 can only be a bare,
+	// un-enveloped empty struct.
+	// 2. A message of length >1 starting with 0x00 can only be enveloped
+	// without a version (non-strict)
```
can you provide rationale here. e.g., why can't it be a valid struct (since field type starts a struct, and 0 is not a valid field type).

prashantv commented on protocol/binary.go:
```diff
@@ -61,3 +82,108 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
+}
+
+// noEnvelopeResponder responds to a request without an envelope.
+
```
nit: comment should be right above the declaration.

prashantv commented on protocol/binary.go:
```diff
@@ -61,3 +82,108 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
+
+type _noEnvelopeResponder struct{}
+
+func (_noEnvelopeResponder) EncodeResponse(v wire.Value, t wire.EnvelopeType, w io.Writer) error {
```
should we go the other way (underscore prefix for the variable, and not the type, since that also matches our internal style guide)

prashantv commented on protocol/binary.go:
```diff
@@ -61,3 +82,108 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
+		}, nil
+	}
+
+	// All other patterns are either bare structs or invalid.
```
we should call out why this isn't filtering on just the `0x00-0x0f` which is what the rule called out. specifically that we're OK with assuming that we don't see large procedure fields?

breerly replied:
+1

prashantv commented on protocol/binary.go:
```diff
@@ -61,3 +82,108 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
+
+var noEnvelopeResponder EnvelopeSpecificResponder = &_noEnvelopeResponder{}
+
+// envelopeV0Responder responds to requests with a non-strict (unversioned) envelope.
```
the comment should go right above the declaration without a blank line.

prashantv commented on protocol/binary/envelope.go:
```diff
@@ -52,6 +51,24 @@ func (bw *Writer) WriteEnveloped(e wire.Envelope) error {
 
+// WriteLegacyEnveloped writes enveloped value using the non-strict envelope
+// (non-strict lacks an envelope version).
+func (bw *Writer) WriteLegacyEnveloped(e wire.Envelope) error {
```
should we export this? can we just keep it private since we're pretty confident there isn't a good reason for users to use this?

kriskowal replied:
In this case, we’re using protocol/binary.Writer from the protocol package, so this does need to be public. We might have had the foresight to put this in the same package or an internal package, but sadly missed that opportunity.

abhinav replied:
I don't feel strongly about this but as discussed offline,
WriteLegacyEnveloped isn't necessary because protocol implementations are
expected to be able to decode both, strict and non-strict envelopes regardless
of what they're sending.

Looking at the Apache Thrift Python library implementation, non-strict
envelopes are rejected only if the user explicitly opted into it. Strict
envelopes (which is what ThriftRW Go uses) are never rejected, even if the
user opted into writing non-strict envelopes.

abhinav replied:
I buy it.

prashantv replied:
The logic to detect the type of envelope (if any) seems very tied to the `binary` protocol -- should all of that logic live inside of the `binary` package? That way `WriteLegacyEnvelope` is not exported, but we still support it for this specific use case.

prashantv commented on protocol/binary_test.go:
```diff
@@ -788,3 +821,219 @@ func TestBinaryEnvelopeSuccessful(t *testing.T) {
 }
+
+func TestReqRes(t *testing.T) {
+
```
nit: delete blank

prashantv commented on protocol/binary_test.go:
```diff
@@ -788,3 +821,219 @@ func TestBinaryEnvelopeSuccessful(t *testing.T) {
+				0x00, 0x01, // id:2 = 1
+				0x00, 0x2a, // value = 42
+
+				0x0F,       // type:1 = list
```
can we avoid manually encoding these bytes, and use the existing binary encoding? prefer to avoid re-testing the struct binary encoding here, which we already test elsewhere. the binary that is relevant here is the envelope, the rest should be `encode(vstruct(..))`

breerly commented on protocol/binary.go:
```diff
@@ -21,17 +21,45 @@
+
 func init() {
 	Binary = binaryProtocol{}
+	EnvelopeAgnosticBinary = binaryProtocol{}
```
Great naming.

breerly commented on protocol/binary.go:
```diff
@@ -21,17 +21,45 @@
 )
 
 // Binary implements the Thrift Binary Protocol.
+// Binary can be cast up to EnvelopeAgnosticProtocol to support DecodeRequest.
```
ty

breerly commented on protocol/binary_test.go:
```diff
@@ -683,16 +684,27 @@ func TestBinaryEnvelopeErrors(t *testing.T) {
 			continue
 		}
 
+		// Also verify that DecodeRequest fails.
```
Thinking outloud - should these be in their own test?

breerly commented on protocol/binary_test.go:
```diff
@@ -776,6 +791,24 @@ func TestBinaryEnvelopeSuccessful(t *testing.T) {
 			continue
 		}
 
+		// Also verify whether we can infer the presence of the envelope
```
Same thing - wondering if the envelope agnostic items should be in their own test.

breerly commented on protocol/binary_test.go:
```diff
@@ -788,3 +821,219 @@ func TestBinaryEnvelopeSuccessful(t *testing.T) {
 	}
 }
+
+func TestReqRes(t *testing.T) {
```
Suggest renaming to describe the behavior being verified... perhaps `TestEnvelopeAgnosticism`...

breerly commented on protocol/protocol.go:
```diff
@@ -44,3 +44,28 @@ type Protocol interface {
+// EnvelopeAgnosticProtocol defines a specific way for a Thrift value to be
+// encoded or decoded, additionally being able to decode requests without prior
+// knowledge of whether the request is enveloped.
+type EnvelopeAgnosticProtocol interface {
```
Might be worth also adding a comment about upcasting Protocol to this interface here.

abhinav commented on protocol/binary.go:
```diff
@@ -61,3 +89,116 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
+	// Strip request envelopes if present.
+	// 1. A message of length 1 containing only 0x00 can only be a bare,
+	// un-enveloped empty struct.
+	// 2. A message of length >1 starting with 0x00 can only be enveloped
```
I think the explanation should include the actual expected shapes of the
payloads.

```
Enveloped (strict):
    version:2 0x00 envType:1 name~4 id:4
    where version = 0x8001
Enveloped (non-strict):
    name~4 envType:1 id:4
Non-enveloped:
    fieldType:1 (fieldID:2 fieldData:? fieldType:1)*
    when fieldType = 0, no more information follows
```

Plus it can be more obvious.

```
A message can be,

1.  Enveloped (strict)
    
    The message must begin with 0x80 -- the first byte of the version number.

2.  Enveloped (non-strict)

    The message begins with the length of the method name. As long as the
    method name is not 16 MB long, the first byte of this message is 0x00. If
    the message contains at least 9 other bytes, it is enveloped using the
    non-strict format.

        4  bytes     message length (minimum = 0)
        1  byte      message type
        4  bytes     sequence ID
        1  byte      empty struct (0x00)
        --------
        10 bytes
        
3.  Unenveloped

    If the message begins with 0x00 but does not contain any more bytes, it's
    a bare, un-enveloped empty struct.

    If the message begins with any other byte, it's an unenveloped message or
    an invalid request.
```

abhinav commented on protocol/binary.go:
```diff
@@ -61,3 +89,116 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
+	// beginning of an un-enveloped field.
+	var buf [2]byte
+	// If we fail to read two bytes, the only possible valid value is the empty struct.
+	if _, err := r.ReadAt(buf[0:2], 0); err != nil {
```
Instead of relying on `err != nil`, it may be preferable to rely on the first return value of `ReadAt`. (We're ignoring the error anyway.)

```lang=go
if n, _ := r.ReadAt(buf[:2], 0); n < 2 {
    ...
}
```

abhinav commented on protocol/binary.go:
```diff
@@ -21,17 +21,90 @@
+//      non-strict format.
+//
+//          4  bytes     method name length (minimum = 0)
+//			n  bytes	 method name
```
alignment

abhinav commented on protocol/binary.go:
```diff
@@ -21,17 +21,90 @@
+//      a bare, un-enveloped empty struct.
+//
+//      If the message begins with any other byte, it's an unenveloped message or
+//      an invalid request./
```
trailing `/`

abhinav commented on protocol/binary.go:
```diff
@@ -61,3 +134,133 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
+	var buf [2]byte
+
+	// If we fail to read two bytes, the only possible valid value is the empty struct.
+	if _, err := r.ReadAt(buf[0:2], 0); err != nil {
```
repeating nit because comment was hidden by GH:
Probably better to use the integer return value since there are other reasons
for `err` to be non-nil.

```go
if n, err := r.ReadAt(...); n == 1 {
    // Only empty struct possible.
} else if err != nil {
    // Read error occurred.
    return wire.Value{}, nil, err
}
```

prashantv replied:
`ReadAt` returns a non-nil error if the number of bytes read is `< len(buf)`:
> ReadAt reads len(p) bytes into p starting at offset off in the underlying input source. It returns the number of bytes read (0 <= n <= len(p)) and any error encountered.
>
>When ReadAt returns n < len(p), it returns a non-nil error explaining why more bytes were not returned. In this respect, ReadAt is stricter than Read.

abhinav replied:
`ReadAt` could read zero bytes because of some IO error. That will be treated as a successful read of an empty struct right now.

abhinav commented on protocol/binary.go:
```diff
@@ -61,3 +134,133 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
+	val, err := b.Decode(r, wire.TStruct)
+	if err != nil {
+		return wire.Value{}, _noEnvelopeResponder, err
+	}
```
optional nitty nit: can simplify

    val, err := b.Decode(...)
    return val, _noEnvelopeResponder, err

abhinav commented on protocol/binary_test.go:
```diff
@@ -776,6 +791,24 @@ func TestBinaryEnvelopeSuccessful(t *testing.T) {
+			continue
+		}
+
+		if !assert.True(t, tt.wantResponderType == reflect.TypeOf(responder), "%v: decoded request should have responder want %v got %T", tt.msg, tt.wantResponderType, responder) {
```
optional: can `assert.Equal` on `reflect.Type`s

```go
assert.Equal(t, tt.wantResponderType, reflect.TypeOf(responder))
```

abhinav commented on protocol/binary_test.go:
```diff
@@ -788,3 +821,137 @@ func TestBinaryEnvelopeSuccessful(t *testing.T) {
+}
+
+func TestReqRes(t *testing.T) {
+
```
nit: delete empty line at start of function

I think I read somewhere that upcoming versions of golint will complain about this

abhinav commented on protocol/binary_test.go:
```diff
@@ -788,3 +821,137 @@ func TestBinaryEnvelopeSuccessful(t *testing.T) {
+			continue
+		}
+
+		writer := bytes.NewBuffer(make([]byte, 0, 0))
```
`new(bytes.Buffer)`?

kriskowal replied:
went with &bytes.Buffer{}

prashantv commented on protocol/binary.go:
```diff
@@ -61,3 +134,133 @@ func (binaryProtocol) DecodeEnveloped(r io.ReaderAt) (wire.Envelope, error) {
+		if err != nil {
+			return wire.Value{}, _noEnvelopeResponder, err
+		}
+		return val, _noEnvelopeResponder, nil
```
optional, can this be simplified to
```lang=go
val, err := b.Decode(r, wire.TStruct)
return val, _noEnvelopeResponder, err
```

kriskowal commented:
@prashantv I agree that putting all the binary protocol bits in binary would be good, but I’m making the call to avoid a yak shave into restructuring the existing code.

Taking everything else ;)

abhinav commented on protocol/binary.go:
```diff
@@ -221,15 +221,16 @@ func (noEnvelopeResponder) EncodeResponse(v wire.Value, t wire.EnvelopeType, w i
 
-var _noEnvelopeResponder EnvelopeSpecificResponder = &noEnvelopeResponder{}
+// NoEnvelopeResponder responds to a request without an envelope.
+var NoEnvelopeResponder EnvelopeSpecificResponder = &noEnvelopeResponder{}
```
nit: as discussed offline, I think we should call the interface just Responder
